File: benchmarks/scripts/compare_metrics.py
#!/usr/bin/env python3
"""Compare complexity metrics between ComplexityGuard and FTA for a single project.

Usage:
    python3 compare_metrics.py <cg-json-path> <fta-json-path> <project-name>

Output:
    JSON comparison to stdout.
    Human-readable summary to stderr.

CG JSON schema: { files: [{ path, functions: [{ name, cyclomatic, halstead_volume, ... }] }] }
FTA JSON schema: [{ file_name, cyclo, halstead: { volume, ... }, line_count, ... }]

Methodology:
    CG operates at function-level granularity. To compare with FTA's file-level output,
    we aggregate CG values per file by summing per-function values. This produces comparable
    totals while preserving CG's higher granularity for other analysis purposes.

    Tolerance bands account for parser differences (CG uses tree-sitter, FTA uses SWC)
    and aggregation differences:
      - cyclomatic: 25% tolerance
      - halstead_volume: 30% tolerance (SWC tokenizes differently than tree-sitter)
      - line_count: 20% tolerance (different line counting rules)
"""

import json
import sys
import os
import math


CYCLOMATIC_TOLERANCE = 25.0
HALSTEAD_TOLERANCE = 30.0
LINE_COUNT_TOLERANCE = 20.0


def normalize_cg_path(path: str, project_name: str) -> str:
    """Extract path relative to the project root from an absolute CG file path.

    CG produces absolute paths like:
      /home/user/.../benchmarks/projects/zod/src/types.ts
    FTA produces paths relative to the project root like:
      src/types.ts

    This function strips everything up to and including the project name segment
    so both tools produce comparable relative paths.
    """
    parts = path.replace("\\", "/").split("/")
    try:
        idx = parts.index(project_name)
        return "/".join(parts[idx + 1:])
    except ValueError:
        # If project name not found in path, return as-is (already relative or unknown)
        return path.replace("\\", "/")


def load_cg_output(cg_path: str, project_name: str) -> dict:
    """Load and aggregate CG output to file-level metrics.

    Returns dict mapping relative_path -> { cyclomatic, halstead_volume, line_count }
    """
    with open(cg_path) as f:
        data = json.load(f)

    file_metrics = {}
    for file_entry in data.get("files", []):
        rel_path = normalize_cg_path(file_entry.get("path", ""), project_name)
        functions = file_entry.get("functions", [])

        # Aggregate function-level values to file-level sums
        total_cyclomatic = sum(fn.get("cyclomatic", 0) for fn in functions)
        total_halstead_volume = sum(fn.get("halstead_volume", 0.0) for fn in functions)

        # file_length is a direct file-level field on the CG file entry
        line_count = file_entry.get("file_length", 0)

        if not rel_path:
            continue

        file_metrics[rel_path] = {
            "cyclomatic": total_cyclomatic,
            "halstead_volume": total_halstead_volume,
            "line_count": line_count,
            "function_count": len(functions),
        }

    return file_metrics


def load_fta_output(fta_path: str, project_name: str) -> dict:
    """Load FTA output and return file-level metrics.

    FTA produces paths already relative to the project root (e.g., "src/types.ts").
    No normalization needed beyond forward-slash conversion.

    Returns dict mapping relative_path -> { cyclomatic, halstead_volume, line_count }
    """
    with open(fta_path) as f:
        data = json.load(f)

    file_metrics = {}
    for entry in data:
        # FTA file_name is already relative to project root
        file_name = entry.get("file_name", "").replace("\\", "/")

        halstead = entry.get("halstead", {})
        file_metrics[file_name] = {
            "cyclomatic": entry.get("cyclo", 0),
            "halstead_volume": halstead.get("volume", 0.0),
            "line_count": entry.get("line_count", 0),
        }

    return file_metrics


def diff_pct(cg_val: float, fta_val: float) -> float:
    """Compute normalized percentage difference between two values."""
    denom = max(abs(cg_val), abs(fta_val), 1.0)
    return abs(cg_val - fta_val) / denom * 100.0


def compute_ranking_correlation(cg_metrics: dict, fta_metrics: dict, metric: str,
                                 common_files: list) -> float:
    """Compute Spearman rank correlation for a metric across common files.

    Returns correlation coefficient in [-1, 1].
    """
    if len(common_files) < 2:
        return 0.0

    cg_vals = [cg_metrics[f][metric] for f in common_files]
    fta_vals = [fta_metrics[f][metric] for f in common_files]

    def rank_list(values):
        sorted_vals = sorted(enumerate(values), key=lambda x: x[1])
        ranks = [0.0] * len(values)
        for rank, (orig_idx, _) in enumerate(sorted_vals):
            ranks[orig_idx] = float(rank + 1)
        return ranks

    cg_ranks = rank_list(cg_vals)
    fta_ranks = rank_list(fta_vals)

    n = len(common_files)
    mean_cg = sum(cg_ranks) / n
    mean_fta = sum(fta_ranks) / n

    num = sum((cg_ranks[i] - mean_cg) * (fta_ranks[i] - mean_fta) for i in range(n))
    denom_cg = math.sqrt(sum((r - mean_cg) ** 2 for r in cg_ranks))
    denom_fta = math.sqrt(sum((r - mean_fta) ** 2 for r in fta_ranks))

    if denom_cg < 1e-10 or denom_fta < 1e-10:
        return 0.0

    return num / (denom_cg * denom_fta)


def analyze_metric(cg_metrics: dict, fta_metrics: dict, metric: str,
                   tolerance: float, common_files: list) -> dict:
    """Compute comparison stats for a single metric across all common files."""
    diffs = []
    within_tolerance_count = 0

    for path in common_files:
        cg_val = float(cg_metrics[path].get(metric, 0))
        fta_val = float(fta_metrics[path].get(metric, 0))
        d = diff_pct(cg_val, fta_val)
        diffs.append(d)
        if d <= tolerance:
            within_tolerance_count += 1

    n = len(common_files)
    if n == 0:
        return {
            "within_tolerance_pct": 0.0,
            "mean_diff_pct": 0.0,
            "ranking_correlation": 0.0,
        }

    mean_diff = sum(diffs) / n
    within_pct = within_tolerance_count / n * 100.0
    correlation = compute_ranking_correlation(cg_metrics, fta_metrics, metric, common_files)

    return {
        "within_tolerance_pct": round(within_pct, 1),
        "mean_diff_pct": round(mean_diff, 1),
        "ranking_correlation": round(correlation, 4),
    }


def main():
    if len(sys.argv) != 4:
        print("Usage: compare_metrics.py <cg-json> <fta-json> <project-name>", file=sys.stderr)
        sys.exit(1)

    cg_path = sys.argv[1]
    fta_path = sys.argv[2]
    project_name = sys.argv[3]

    # Load outputs
    try:
        cg_metrics = load_cg_output(cg_path, project_name)
    except (json.JSONDecodeError, FileNotFoundError) as e:
        print(f"Warning: Could not load CG output from {cg_path}: {e}", file=sys.stderr)
        cg_metrics = {}

    try:
        fta_metrics = load_fta_output(fta_path, project_name)
    except (json.JSONDecodeError, FileNotFoundError) as e:
        print(f"Warning: Could not load FTA output from {fta_path}: {e}", file=sys.stderr)
        fta_metrics = {}

    # Find common files
    cg_paths = set(cg_metrics.keys())
    fta_paths = set(fta_metrics.keys())
    common_files = sorted(cg_paths & fta_paths)
    cg_only = len(cg_paths - fta_paths)
    fta_only = len(fta_paths - cg_paths)

    # Analyze each metric
    cyclomatic_stats = analyze_metric(
        cg_metrics, fta_metrics, "cyclomatic", CYCLOMATIC_TOLERANCE, common_files
    )
    halstead_stats = analyze_metric(
        cg_metrics, fta_metrics, "halstead_volume", HALSTEAD_TOLERANCE, common_files
    )
    line_count_stats = analyze_metric(
        cg_metrics, fta_metrics, "line_count", LINE_COUNT_TOLERANCE, common_files
    )

    result = {
        "project": project_name,
        "files_compared": len(common_files),
        "files_cg_only": cg_only,
        "files_fta_only": fta_only,
        "cyclomatic": cyclomatic_stats,
        "halstead_volume": halstead_stats,
        "line_count": line_count_stats,
        "methodology": {
            "cg_aggregation": "sum of per-function values",
            "fta_granularity": "file-level",
            "cyclomatic_tolerance": CYCLOMATIC_TOLERANCE,
            "halstead_tolerance": HALSTEAD_TOLERANCE,
            "line_count_tolerance": LINE_COUNT_TOLERANCE,
            "note": (
                "FTA uses SWC parser; CG uses tree-sitter. "
                "Different tokenization rules cause expected divergence. "
                "Tolerance bands account for parser and aggregation differences."
            ),
        },
    }

    print(json.dumps(result, indent=2))

    # Human-readable summary to stderr
    print(f"\n--- Metric Accuracy: {project_name} ---", file=sys.stderr)
    print(f"Files compared: {len(common_files)} (CG-only: {cg_only}, FTA-only: {fta_only})", file=sys.stderr)
    print(f"Cyclomatic:     {cyclomatic_stats['within_tolerance_pct']:5.1f}% within {CYCLOMATIC_TOLERANCE}% tolerance, "
          f"mean diff {cyclomatic_stats['mean_diff_pct']:.1f}%, "
          f"rank corr {cyclomatic_stats['ranking_correlation']:.3f}", file=sys.stderr)
    print(f"Halstead vol:   {halstead_stats['within_tolerance_pct']:5.1f}% within {HALSTEAD_TOLERANCE}% tolerance, "
          f"mean diff {halstead_stats['mean_diff_pct']:.1f}%, "
          f"rank corr {halstead_stats['ranking_correlation']:.3f}", file=sys.stderr)
    print(f"Line count:     {line_count_stats['within_tolerance_pct']:5.1f}% within {LINE_COUNT_TOLERANCE}% tolerance, "
          f"mean diff {line_count_stats['mean_diff_pct']:.1f}%, "
          f"rank corr {line_count_stats['ranking_correlation']:.3f}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: zig/benchmarks/scripts/compare-metrics.mjs
#!/usr/bin/env node
/**
 * Compare complexity metrics between ComplexityGuard and FTA for a single project.
 *
 * Usage:
 *   node compare-metrics.mjs <cg-json-path> <fta-json-path> <project-name>
 *
 * Output:
 *   JSON comparison to stdout.
 *   Human-readable summary to stderr.
 *
 * CG JSON schema: { files: [{ path, functions: [{ name, cyclomatic, halstead_volume, ... }] }] }
 * FTA JSON schema: [{ file_name, cyclo, halstead: { volume, ... }, line_count, ... }]
 *
 * Methodology:
 *   CG operates at function-level granularity. To compare with FTA's file-level output,
 *   we aggregate CG values per file by summing per-function values. This produces comparable
 *   totals while preserving CG's higher granularity for other analysis purposes.
 *
 *   Tolerance bands account for parser differences (CG uses tree-sitter, FTA uses SWC)
 *   and aggregation differences:
 *     - cyclomatic: 25% tolerance
 *     - halstead_volume: 30% tolerance (SWC tokenizes differently than tree-sitter)
 *     - line_count: 20% tolerance (different line counting rules)
 */

import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';

const CYCLOMATIC_TOLERANCE = 25.0;
const HALSTEAD_TOLERANCE = 30.0;
const LINE_COUNT_TOLERANCE = 20.0;

/**
 * Extract path relative to the project root from an absolute CG file path.
 *
 * CG produces absolute paths like:
 *   /home/user/.../benchmarks/projects/zod/src/types.ts
 * FTA produces paths relative to the project root like:
 *   src/types.ts
 *
 * This function strips everything up to and including the project name segment
 * so both tools produce comparable relative paths.
 * @param {string} filePath
 * @param {string} projectName
 * @returns {string}
 */
function normalizeCgPath(filePath, projectName) {
  const normalized = filePath.replace(/\\/g, '/');
  const parts = normalized.split('/');
  const idx = parts.indexOf(projectName);
  if (idx === -1) {
    // If project name not found in path, return as-is (already relative or unknown)
    return normalized;
  }
  return parts.slice(idx + 1).join('/');
}

/**
 * Load and aggregate CG output to file-level metrics.
 *
 * Returns object mapping relative_path -> { cyclomatic, halstead_volume, line_count, function_count }
 * @param {string} cgPath
 * @param {string} projectName
 * @returns {object}
 */
function loadCgOutput(cgPath, projectName) {
  const content = fs.readFileSync(cgPath, 'utf8');
  const data = JSON.parse(content);

  const fileMetrics = {};
  for (const fileEntry of data.files || []) {
    const relPath = normalizeCgPath(fileEntry.path || '', projectName);
    const functions = fileEntry.functions || [];

    // Aggregate function-level values to file-level sums
    const totalCyclomatic = functions.reduce((sum, fn) => sum + (fn.cyclomatic || 0), 0);
    const totalHalsteadVolume = functions.reduce((sum, fn) => sum + (fn.halstead_volume || 0.0), 0);

    // file_length is a direct file-level field on the CG file entry
    const lineCount = fileEntry.file_length || 0;

    if (!relPath) continue;

    fileMetrics[relPath] = {
      cyclomatic: totalCyclomatic,
      halstead_volume: totalHalsteadVolume,
      line_count: lineCount,
      function_count: functions.length,
    };
  }

  return fileMetrics;
}

/**
 * Load FTA output and return file-level metrics.
 *
 * FTA produces paths already relative to the project root (e.g., "src/types.ts").
 * No normalization needed beyond forward-slash conversion.
 *
 * Returns object mapping relative_path -> { cyclomatic, halstead_volume, line_count }
 * @param {string} ftaPath
 * @param {string} _projectName
 * @returns {object}
 */
function loadFtaOutput(ftaPath, _projectName) {
  const content = fs.readFileSync(ftaPath, 'utf8');
  const data = JSON.parse(content);

  const fileMetrics = {};
  for (const entry of data) {
    const fileName = (entry.file_name || '').replace(/\\/g, '/');
    const halstead = entry.halstead || {};
    fileMetrics[fileName] = {
      cyclomatic: entry.cyclo || 0,
      halstead_volume: halstead.volume || 0.0,
      line_count: entry.line_count || 0,
    };
  }

  return fileMetrics;
}

/**
 * Compute normalized percentage difference between two values.
 * @param {number} cgVal
 * @param {number} ftaVal
 * @returns {number}
 */
function diffPct(cgVal, ftaVal) {
  const denom = Math.max(Math.abs(cgVal), Math.abs(ftaVal), 1.0);
  return (Math.abs(cgVal - ftaVal) / denom) * 100.0;
}

/**
 * Compute Spearman rank correlation for a metric across common files.
 *
 * Returns correlation coefficient in [-1, 1].
 * @param {object} cgMetrics
 * @param {object} ftaMetrics
 * @param {string} metric
 * @param {string[]} commonFiles
 * @returns {number}
 */
function computeRankingCorrelation(cgMetrics, ftaMetrics, metric, commonFiles) {
  if (commonFiles.length < 2) return 0.0;

  const cgVals = commonFiles.map(f => cgMetrics[f][metric]);
  const ftaVals = commonFiles.map(f => ftaMetrics[f][metric]);

  function rankList(values) {
    const indexed = values.map((v, i) => [i, v]);
    indexed.sort((a, b) => a[1] - b[1]);
    const ranks = new Array(values.length).fill(0.0);
    for (let rank = 0; rank < indexed.length; rank++) {
      ranks[indexed[rank][0]] = rank + 1;
    }
    return ranks;
  }

  const cgRanks = rankList(cgVals);
  const ftaRanks = rankList(ftaVals);

  const n = commonFiles.length;
  const meanCg = cgRanks.reduce((sum, r) => sum + r, 0) / n;
  const meanFta = ftaRanks.reduce((sum, r) => sum + r, 0) / n;

  let num = 0;
  for (let i = 0; i < n; i++) {
    num += (cgRanks[i] - meanCg) * (ftaRanks[i] - meanFta);
  }

  const denomCg = Math.sqrt(cgRanks.reduce((sum, r) => sum + (r - meanCg) ** 2, 0));
  const denomFta = Math.sqrt(ftaRanks.reduce((sum, r) => sum + (r - meanFta) ** 2, 0));

  if (denomCg < 1e-10 || denomFta < 1e-10) return 0.0;

  return num / (denomCg * denomFta);
}

/**
 * Compute comparison stats for a single metric across all common files.
 * @param {object} cgMetrics
 * @param {object} ftaMetrics
 * @param {string} metric
 * @param {number} tolerance
 * @param {string[]} commonFiles
 * @returns {object}
 */
function analyzeMetric(cgMetrics, ftaMetrics, metric, tolerance, commonFiles) {
  const diffs = [];
  let withinToleranceCount = 0;

  for (const filePath of commonFiles) {
    const cgVal = Number(cgMetrics[filePath][metric] || 0);
    const ftaVal = Number(ftaMetrics[filePath][metric] || 0);
    const d = diffPct(cgVal, ftaVal);
    diffs.push(d);
    if (d <= tolerance) withinToleranceCount++;
  }

  const n = commonFiles.length;
  if (n === 0) {
    return {
      within_tolerance_pct: 0.0,
      mean_diff_pct: 0.0,
      ranking_correlation: 0.0,
    };
  }

  const meanDiff = diffs.reduce((sum, d) => sum + d, 0) / n;
  const withinPct = (withinToleranceCount / n) * 100.0;
  const correlation = computeRankingCorrelation(cgMetrics, ftaMetrics, metric, commonFiles);

  return {
    within_tolerance_pct: Math.round(withinPct * 10) / 10,
    mean_diff_pct: Math.round(meanDiff * 10) / 10,
    ranking_correlation: Math.round(correlation * 10000) / 10000,
  };
}

function main() {
  const args = process.argv.slice(2);

  if (args.length !== 3) {
    process.stderr.write('Usage: compare-metrics.mjs <cg-json> <fta-json> <project-name>\n');
    process.exit(1);
  }

  const cgPath = args[0];
  const ftaPath = args[1];
  const projectName = args[2];

  // Load outputs
  let cgMetrics = {};
  try {
    cgMetrics = loadCgOutput(cgPath, projectName);
  } catch (e) {
    process.stderr.write(`Warning: Could not load CG output from ${cgPath}: ${e.message}\n`);
  }

  let ftaMetrics = {};
  try {
    ftaMetrics = loadFtaOutput(ftaPath, projectName);
  } catch (e) {
    process.stderr.write(`Warning: Could not load FTA output from ${ftaPath}: ${e.message}\n`);
  }

  // Find common files
  const cgPaths = new Set(Object.keys(cgMetrics));
  const ftaPaths = new Set(Object.keys(ftaMetrics));
  const commonFiles = [...cgPaths].filter(f => ftaPaths.has(f)).sort();
  const cgOnly = cgPaths.size - commonFiles.length;
  const ftaOnly = ftaPaths.size - commonFiles.length;

  // Analyze each metric
  const cyclomaticStats = analyzeMetric(cgMetrics, ftaMetrics, 'cyclomatic', CYCLOMATIC_TOLERANCE, commonFiles);
  const halsteadStats = analyzeMetric(cgMetrics, ftaMetrics, 'halstead_volume', HALSTEAD_TOLERANCE, commonFiles);
  const lineCountStats = analyzeMetric(cgMetrics, ftaMetrics, 'line_count', LINE_COUNT_TOLERANCE, commonFiles);

  const result = {
    project: projectName,
    files_compared: commonFiles.length,
    files_cg_only: cgOnly,
    files_fta_only: ftaOnly,
    cyclomatic: cyclomaticStats,
    halstead_volume: halsteadStats,
    line_count: lineCountStats,
    methodology: {
      cg_aggregation: 'sum of per-function values',
      fta_granularity: 'file-level',
      cyclomatic_tolerance: CYCLOMATIC_TOLERANCE,
      halstead_tolerance: HALSTEAD_TOLERANCE,
      line_count_tolerance: LINE_COUNT_TOLERANCE,
      note:
        'FTA uses SWC parser; CG uses tree-sitter. ' +
        'Different tokenization rules cause expected divergence. ' +
        'Tolerance bands account for parser and aggregation differences.',
    },
  };

  process.stdout.write(JSON.stringify(result, null, 2) + '\n');

  // Human-readable summary to stderr
  process.stderr.write(`\n--- Metric Accuracy: ${projectName} ---\n`);
  process.stderr.write(`Files compared: ${commonFiles.length} (CG-only: ${cgOnly}, FTA-only: ${ftaOnly})\n`);
  process.stderr.write(
    `Cyclomatic:     ${cyclomaticStats.within_tolerance_pct.toFixed(1).padStart(5)}% within ${CYCLOMATIC_TOLERANCE}% tolerance, ` +
    `mean diff ${cyclomaticStats.mean_diff_pct.toFixed(1)}%, ` +
    `rank corr ${cyclomaticStats.ranking_correlation.toFixed(3)}\n`
  );
  process.stderr.write(
    `Halstead vol:   ${halsteadStats.within_tolerance_pct.toFixed(1).padStart(5)}% within ${HALSTEAD_TOLERANCE}% tolerance, ` +
    `mean diff ${halsteadStats.mean_diff_pct.toFixed(1)}%, ` +
    `rank corr ${halsteadStats.ranking_correlation.toFixed(3)}\n`
  );
  process.stderr.write(
    `Line count:     ${lineCountStats.within_tolerance_pct.toFixed(1).padStart(5)}% within ${LINE_COUNT_TOLERANCE}% tolerance, ` +
    `mean diff ${lineCountStats.mean_diff_pct.toFixed(1)}%, ` +
    `rank corr ${lineCountStats.ranking_correlation.toFixed(3)}\n`
  );
}

main();
